packages/x-data-grid/src/hooks/core/useGridVirtualizer.tsx

```diff
@@ -0,0 +1,296 @@
+import * as React from 'react';
+import useEventCallback from '@mui/utils/useEventCallback';
+import { useRtl } from '@mui/system/RtlProvider';
+import { RefObject } from '@mui/x-internals/types';
+import { roundToDecimalPlaces } from '@mui/x-internals/math';
+import { useVirtualizer } from '@mui/x-virtualizer';
+import { useFirstRender } from '../utils/useFirstRender';
+import { GridPrivateApiCommunity } from '../../models/api/gridApiCommunity';
+import { GridStateColDef } from '../../models/colDef/gridColDef';
+import { createSelector } from '../../utils/createSelector';
+import { useGridSelector } from '../utils/useGridSelector';
+import {
+  gridContentHeightSelector,
+  gridHasFillerSelector,
+  gridVerticalScrollbarWidthSelector,
+} from '../features/dimensions/gridDimensionsSelectors';
+import { gridDensityFactorSelector } from '../features/density';
+import {
+  gridVisibleColumnDefinitionsSelector,
+  gridVisiblePinnedColumnDefinitionsSelector,
+  gridColumnPositionsSelector,
+  gridHasColSpanSelector,
+} from '../features/columns/gridColumnsSelector';
+import { gridPinnedRowsSelector, gridRowCountSelector } from '../features/rows/gridRowsSelector';
+import { useGridVisibleRows } from '../utils/useGridVisibleRows';
+import { DataGridProcessedProps } from '../../models/props/DataGridProps';
+import { gridPaginationSelector } from '../features/pagination';
+import { gridListColumnSelector } from '../features/listView/gridListViewSelectors';
+import { minimalContentHeight } from '../features/rows/gridRowsUtils';
+import { EMPTY_PINNED_COLUMN_FIELDS, GridPinnedColumns } from '../features/columns';
+import { gridFocusedVirtualCellSelector } from '../features/virtualization/gridFocusedVirtualCellSelector';
+import { gridRowSelectionManagerSelector } from '../features/rowSelection';
+import { DATA_GRID_PROPS_DEFAULT_VALUES } from '../../constants/dataGridPropsDefaultValues';
+import { getValidRowHeight, rowHeightWarning } from '../features/rows/gridRowsUtils';
+import { getTotalHeaderHeight } from '../features/columns/gridColumnsUtils';
+
+type RootProps = DataGridProcessedProps;
+
+const columnsTotalWidthSelector = createSelector(
+  gridVisibleColumnDefinitionsSelector,
+  gridColumnPositionsSelector,
+  (visibleColumns, positions) => {
+    const colCount = visibleColumns.length;
+    if (colCount === 0) {
+      return 0;
+    }
+    return roundToDecimalPlaces(
+      positions[colCount - 1] + visibleColumns[colCount - 1].computedWidth,
+      1,
+    );
+  },
+);
+
+/**
+ * Virtualizer setup
+ */
+export function useGridVirtualizer(
+  apiRef: RefObject<GridPrivateApiCommunity>,
+  rootProps: RootProps,
+): void {
+  const isRtl = useRtl();
+  const { listView } = rootProps;
+  const visibleColumns = useGridSelector(apiRef, () => {
+    if (listView) {
+      const column = gridListColumnSelector(apiRef);
+      return column ? [column] : [];
+    }
+    return gridVisibleColumnDefinitionsSelector(apiRef);
+  });
```

Comment:
@KenanYusuf As I mentioned [back here](https://github.com/mui/mui-x/pull/14393#discussion_r1802624870), it would be nice to integrate list-view to the `gridVisibleColumnDefinitionsSelector`, it would avoid having to repeat this snippet of code everywhere we need the visible columns.

Reply:
Started working on this here https://github.com/mui/mui-x/pull/18724